### ISSUE TITLE
Dealing with strings as iterables

### ISSUE THREAD

**Issue by tbekolay** [nengo/nengo#949]
Currently, we have lots of checks that use `is_iterable`. In most cases, we do this so that code works properly whether the user gives us a `list`, `tuple`, `set`, or some other collection. However, there are lots of iterables out there, including strings, which can easily be accidentally provided (e.g., if using positional instead of keyword arguments). Right now, these cases will likely fail at some point, but we can't really say where.

The question is: should we care? Python does not, and in some cases that's helpful. I'm not sure it's ever helpful for us. Of course, this is perhaps also true with other collections, so I'm not sure if strings are worth singling out.

Anyone have strong opinions either way?

**Comment by drasmuss** [comment]
I'd lean towards `is_iterable(string)` returning false.  I think it's more likely that someone passes a string by accident than intentionally as an iterable, and if they really do want to do that all they have to do is pass `list(string)` instead.

**Comment by jgosmann** [comment]
> I'd lean towards is_iterable(string) returning false.

Can we find a better name for `is_iterable` in that case? Because technically a string is iterable ...

**Comment by hunse** [comment]
I agree with @jgosmann. In most cases that I can think of, the iterables that we're targeting are really just lists and generators, so maybe something specific to that? `is_list_like` kinda sounds like our `is_array_like`, making it easier to remember, but it might be a bit vague. It's the best I can think of right now, though.

**Comment by tbekolay** [comment]
I agree with all this as well -- and I kind of like `is_list_like`, makes sense upon seeing it, and isn't too long. There are also cases where we are actually checking for iterables, so having a separate `is_list_like` makes sense to me!

**Comment by tbekolay** [comment]
Adding a note here so we don't forget it.

- [ ] Use `is_list_like` in `TupleParam` so that strings aren't okay (see discussion in 29b7c74a9fdcb2b39c05e61126a8645f57aaea3e).
